### DISCUSSION TITLE
Uart multiplying messages when configuration is changed

### DISCUSSION THREAD

**Discussion by yamenk-gribaudo** [micropython/micropython#12231] [ESP32]
Uart sending previously sended messages when running `init()` with another rx-tx configuration. Minimal example:

```
from machine import UART
import time

uart = UART(1, 115200, bits=8, parity=None, stop=1, timeout=40)

while True:
    print("working")
    time.sleep_ms(100)
    uart.init(rx=4, tx=15)
    time.sleep_ms(100)
    uart.write("uart 1\n")
    
    time.sleep_ms(100)
    uart.init(rx=16, tx=17)
    time.sleep_ms(100)
    uart.write("uart 2\n")
    
    time.sleep_ms(100)
    uart.init(rx=18, tx=5)
    time.sleep_ms(100)
    uart.write("uart 3\n")
```

I expect that if I connect something to the uart 1, I should revieve something like: 
```
uart 1
uart 1
uart 1
...
```

But instead I see:
```
uart 1
uart 2
uart 3
uart 1
...
```

I also try to flush, deinit and redefine the uart each time, but got the same problem. So it seems a bug in the micropython uart implementation.

I could work on it if someone points me in the right direction.
Thanks,

**Comment by robert-hh** [comment]
You did not mention the board. Is it a ESP32?

**Comment by robert-hh** [comment]
What do you get at the other pin's you have chose for output. You might get the same. To stop the output at a TX pins and to resolve a possible conflict at the input pins, you have to disable the rx, tx pins switching input and output to another pin pair. Just declare them as GPIO input (rx) and output(tx) pins using machine.Pin. Set the tx to high.

**Comment by yamenk-gribaudo** [comment]
> You did not mention the board. Is it a ESP32?

Sorry, the board is generic esp32 like [this one](https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRTUcd-R-QNX-VGlXa8EuGdcJjtVo9WX3Y2t2hqKRlt3wW1vJqRxZVHV14VcQ4zzCH7MW0&usqp=CAU)

**Comment by yamenk-gribaudo** [comment]
Thanks for the quick response,

> What do you get at the other pin's you have chose for output. You might get the same. 

I get the same, yes. Its sending all the messages to all devices

>To stop the output at a TX pins and to resolve a possible conflict at the input pins, you have to disable the rx, tx pins switching input and output to another pin pair. Just declare them as GPIO input (rx) and output(tx) pins using machine.Pin. Set the tx to high.

I will try that and post my result here.

**Comment by yamenk-gribaudo** [comment]
> To stop the output at a TX pins and to resolve a possible conflict at the input pins, you have to disable the rx, tx pins switching input and output to another pin pair. Just declare them as GPIO input (rx) and output(tx) pins using machine.Pin. Set the tx to high.

wow, this worked. May I ask what was happening?  To my understanding, only the hardware UARTs have the capacity to save message. Where are the messages being saved after I deinit the uart?

**Comment by robert-hh** [comment]
The UARt does not store any messages. It's more simple. The ESP32 has an internal kind of switchboard, which connects the signals of modules like UART or GPIO to the pins of the chip. You can connect e.g. the UART tx signal to several pins. That happens here. Connecting tx to pin 17 does not remove the previously made connection to pin 15. That is done by connection Pin15 to GPIO.

**Comment by yamenk-gribaudo** [comment]
Thanks, very clear. For the ones ending here in the future, I end up with something like this:
```
def send_msg(rx, tx, msg):
    uart.init(rx=rx, tx=tx)
    uart.write(str(msg) + "\n")
    while uart.txdone() is False:
        time.sleep_ms(1)
    Pin(tx, Pin.OUT)
```

**Comment by robert-hh** [comment]
Do not forget to detach the rx pin. While it seems fine to connect UART TX to several pins, connection RX to several pins leads to an undefined state, where several external outputs are connected together.

**Comment by yamenk-gribaudo** [comment]
Something like this?
```
def send_msg(rx, tx, msg):
    uart.init(rx=rx, tx=tx)
    uart.write(str(msg) + "\n")
    while uart.txdone() is False:
        time.sleep_ms(1)
    Pin(tx, Pin.OUT)
    Pin(rx, Pin.OUT)
```

**Comment by robert-hh** [comment]
```py
    Pin(tx, Pin.OUT, value=1)
    Pin(rx, Pin.IN)
```
Otherwise the output of the ESP32 would drive against the output of the connected serial device. And value=1 ensures, that the connected RX does not see a `break` state.
